File: cases/trees/api-be/services/be-security.weak-crypto.py
# security/weak-crypto (Python lane) — MUST FIRE on the hashlib.md5 line (EXPECTED py:15). See the
# TypeScript sibling (be-security.weak-crypto.ts) for the history: written as a labeled gap while
# the rule read Java only, promoted to a scored expectation on 2026-08-09 when the six-language
# `hash-call` migration landed (this lane's algorithm-named-constructor spelling included).
#
# bad: MD5 over an opaque API token, to key a rate-limit bucket without storing the raw value. The
# fingerprint is short-lived and never a credential itself, so `security/weak-password-hash` — which
# requires a credential word on the call line — stays silent; weak-crypto is the rule that judges it.
# MD5 is still wrong here: a collision lets two distinct callers share one rate-limit bucket.
# good: SHA-256, the same construction with a strong digest — must stay silent.
import hashlib


def bucket_fingerprint(raw_token: str) -> str:
    return hashlib.md5(raw_token.encode()).hexdigest()[:16]


def good_bucket_fingerprint(raw_token: str) -> str:
    return hashlib.sha256(raw_token.encode()).hexdigest()[:16]

File: cases/trees/api-be/services/be-security.weak-crypto.ts
// security/weak-crypto (TypeScript lane) — MUST FIRE on the createHash('md5') line (EXPECTED ts:23).
//
// HISTORY: written BEFORE the multi-language migration as a labeled gap — the rule then read Java
// only, so this file was expected absent and its line lived in EXPECTED.jsonc's `gap` object. The
// migration landed 2026-08-09: weak-crypto now rides the parser's `hash-call` call-site channel
// across six languages, all three sibling lanes (.ts/.py/.go) fired on the first post-migration
// run, and the entries were promoted to ordinary scored must-fire expectations (`gap` is {} again).
//
// bad: MD5 over a cache key. Not a password hash — `security/weak-password-hash` requires a
// credential word on the call line and correctly stays silent here; judging this non-credential
// class of use is exactly what the migration bought. MD5's collision weakness matters for a cache
// key too: two different requests can be made to collide onto one entry, which is a cache-poisoning
// primitive.
// good: SHA-256, the same call shape with a strong algorithm — must stay silent.

import { createHash } from 'node:crypto';

export function cacheKeyForRoute(route: string, params: Record<string, string>): string {
  const canonical = `${route}?${Object.keys(params)
    .sort()
    .map((k) => `${k}=${params[k]}`)
    .join('&')}`;
  return createHash('md5').update(canonical).digest('hex');
}

export function goodCacheKeyForRoute(route: string, params: Record<string, string>): string {
  const canonical = `${route}?${Object.keys(params)
    .sort()
    .map((k) => `${k}=${params[k]}`)
    .join('&')}`;
  return createHash('sha256').update(canonical).digest('hex');
}
